src/Symfony/Component/Security/Http/Authenticator/Passport/PassportInterface.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Http\Authenticator\Passport;
+
+use Symfony\Component\Security\Core\Exception\BadCredentialsException;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\BadgeInterface;
+
+/**
+ * A Passport contains all security-related information that needs to be
+ * validated during authentication.
+ *
+ * A passport badge can be used to add any additional information to the
+ * passport.
+ *
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ *
+ * @experimental in 5.1
+ */
+interface PassportInterface
+{
+    /**
+     * Adds a new security badge.
+     *
+     * A passport can hold only one instance of the same security badge.
+     * This method replaces the current badge if it is already set on this
+     * passport.
+     *
+     * @return $this
+     */
+    public function addBadge(BadgeInterface $badge): self;
+
+    public function hasBadge(string $badgeFqcn): bool;
+
+    public function getBadge(string $badgeFqcn): ?BadgeInterface;
+
+    /**
+     * Checks if all badges are marked as resolved.
+     *
+     * This makes sure that a CsrfTokenBadge is actually checked and
+     * not ignored.
+     *
+     * @throws BadCredentialsException when a badge is not marked as resolved
```

Comment:
What was the reason to throw this here and return `:void` versus returning a `:bool` and throwing this from `AuthenticatorManager`?

Reply:
A passport can have more than one badge (probably always has more than 1 badge). Doing the check in the passport allows us to have a more expliciet exception message (telling which badge is not resolved).

Another solution would be to do this check completely in `AuthenticatorManager`, but it made more sense to me if a passport can check itself if it's valid.